src/test/objectstore_perf_bench.cc

```diff
@@ -0,0 +1,459 @@
+      int length = block_size;
+     
+      int r = os->read(cid, *oidp, offset, length, data); 
+      assert(r > 0 && r == length);
+    }
+
+    count += read_ops + write_ops;
+    if ((count > 1000) == 0) {
+      __sync_fetch_and_add(&total_ops, count);
+      __sync_fetch_and_add(&total_read_ops, read_ops);
+      __sync_fetch_and_add(&total_write_ops, write_ops);
+      count = 0;
+    }
+  }
+
+   dout(10) << "Thread id " << _my_tid << " done. exiting. " << dendl; 
+}
+
+int main(int argc, const char *argv[])
+{
+
+  // command-line arguments
+  vector<const char*> args;
+  argv_to_vec(argc, argv, args);
+  env_to_vec(args);
+  Config cfg;
+
+  _tid_count = 0;
+  cfg.time_secs = 900;
+  cfg.write_pct = 100;
+  _total_data_written = 0;
+
+  auto cct = global_init(nullptr, args, CEPH_ENTITY_TYPE_OSD,
+			 CODE_ENVIRONMENT_UTILITY, 0);
+
+  std::string val;
+  vector<const char*>::iterator i = args.begin();
+  while (i != args.end()) {
+    if (ceph_argparse_double_dash(args, i))
+      break;
+
+    if (ceph_argparse_witharg(args, i, &val, "--size", (char*)nullptr)) {
+      std::string err;
+      if (!cfg.size.parse(val, &err)) {
+        derr << "error parsing size: " << err << dendl;
+        usage();
+      }
+    } else if (ceph_argparse_witharg(args, i, &val, "--block-size", (char*)nullptr)) {
+      std::string err;
+      if (!cfg.block_size.parse(val, &err)) {
+        derr << "error parsing block-size: " << err << dendl;
+        usage();
+      }
+    } else if (ceph_argparse_witharg(args, i, &val, "--time", (char*)nullptr)) {
+      cfg.time_secs = atoi(val.c_str());
+    } else if (ceph_argparse_witharg(args, i, &val, "--write_pct", (char*)nullptr)) {
+      cfg.write_pct = atoi(val.c_str());
+      if (cfg.write_pct > 100) {
+        derr << "Invalid write precentage ratio. "<< dendl;
+        return -1;
+      }
+    } else if (ceph_argparse_witharg(args, i, &val, "--threads", (char*)nullptr)) {
+      cfg.threads = atoi(val.c_str());
+    } else if (ceph_argparse_witharg(args, i, &val, "--stats_interval", (char*)nullptr)) {
+      cfg.stats_interval = atoi(val.c_str());
+    } else if (ceph_argparse_flag(args, i, "--fill", (char*)nullptr)) {
+      cfg.fill = true;
+    } else {
+      derr << "Error: can't understand argument: " << *i << "\n" << dendl;
+      usage();
+    }
+  }
+
+  common_init_finish(g_ceph_context);
+
+  // create object store
+  dout(0) << "objectstore " << g_conf->osd_objectstore << dendl;
+  dout(0) << "data " << g_conf->osd_data << dendl;
+  dout(0) << "journal " << g_conf->osd_journal << dendl;
+  dout(0) << "size " << cfg.size << dendl;
+  dout(0) << "block-size " << cfg.block_size << dendl;
+  dout(0) << "write% " << cfg.write_pct<< dendl;
+  dout(0) << "fill" << cfg.fill<< dendl;
+  dout(0) << "threads " << cfg.threads << dendl;
+  dout(0) << "stats-interval " << cfg.stats_interval<< dendl;
+  dout(0) << "time " << cfg.time_secs << dendl;
+
+  auto os = std::unique_ptr<ObjectStore>(
```

Comment:
You should probably set cache shards ( via set_cache_shards method call) for the newly created OS similar as OSD and fio plugin do. Otherwise one will have objectstore cache as a single contention point.